grug/app/src/vm.rs

```diff
@@ -266,27 +218,25 @@ where
     AppError: From<VM::Error>,
 {
     // Load the program code from storage and deserialize
-    let code = CODES.load(&storage, code_hash)?;
+    let code = CODES.load(&ctx.storage, code_hash)?;
 
     // Create the providers
-    let querier = QuerierProvider::new(vm.clone(), storage.clone(), gas_tracker.clone(), block);
-    let storage = StorageProvider::new(storage, &[CONTRACT_NAMESPACE, &contract]);
+    let querier = QuerierProvider::new(ctx.clone());
+    let storage = StorageProvider::new(ctx.storage.clone(), &[CONTRACT_NAMESPACE, &contract]);
 
-    Ok(vm.build_instance(
+    Ok(ctx.vm().clone().build_instance(
```

Comment:
No need to clone the VM here. You need a `&mut VM`, just take the `ctx` parameter as mut.